mathics/builtin/numeric.py

```diff
@@ -890,3 +894,87 @@ def compute(user_hash, py_hashtype):
     def apply(self, expr, hashtype, evaluation):
         'Hash[expr_, hashtype_String]'
         return Hash.compute(expr.user_hash, hashtype.get_string_value())
+
+
+class TypeEscalation(Exception):
+    def __init__(self, mode):
+        self.mode = mode
+
+
+class Fold(object):
+    # allows inherited classes to specify a single algorithm implementation that
+    # can be called with machine precision, arbitrary precision or symbolically.
+
+    ComputationFunctions = namedtuple(
+        'ComputationFunctions', ('sin', 'cos'))
+
+    FLOAT = 0
+    MPMATH = 1
+    SYMBOLIC = 2
+
+    math = {
+        FLOAT: ComputationFunctions(
+            cos=math.cos,
+            sin=math.sin,
+        ),
+        MPMATH: ComputationFunctions(
+            cos=mpmath.cos,
+            sin=mpmath.sin,
+        ),
+        SYMBOLIC: ComputationFunctions(
+            cos=lambda x: Expression('Cos', x),
+            sin=lambda x: Expression('Sin', x),
+        )
+    }
+
+    operands = {
+        FLOAT: lambda x: x.to_python(),
+        MPMATH: lambda x: x.to_mpmath(),
```

Comment:
mpmath is a bit finicky; doing this conversion outside a `with mpmath.workprec` block might cause a loss of precision. We should at least add a test against this in `AnglePath`.